contracts/contracts/coordination/SigningCoordinator.sol

```diff
@@ -0,0 +1,379 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import "@openzeppelin-upgradeable/contracts/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/proxy/utils/Initializable.sol";
+import "./ISigningCoordinatorChild.sol";
+import "./ThresholdSigningMultisigCloneFactory.sol";
+import "./SigningCoordinatorDispatcher.sol";
+import "../TACoApplication.sol";
+
+// SigningCoordinator ----> Dispatcher ----> (Relevant) L1Sender ---------[BRIDGE]---------- L2Receiver ----> SigningCoordinatorChild (1. deploys multisig OR 2. updates multisig)
+
+contract SigningCoordinator is Initializable, AccessControlDefaultAdminRulesUpgradeable {
+    using MessageHashUtils for bytes32;
+    using ECDSA for bytes32;
+
+    // Signing protocol
+    event InitiateSigningCohort(
+        uint32 indexed cohortId,
+        uint256 chainId,
+        address indexed authority,
+        address[] participants
+    );
+    event SigningCohortSignaturePosted(
+        uint32 indexed cohortId,
+        address indexed provider,
+        bytes signature
+    );
+    event SigningCohortDeployed(uint32 indexed cohortId, uint256 chainId);
+
+    // Cohort Administration
+    event SigningCohortConditionsSet(uint32 indexed cohortId, uint256 chainId, bytes conditions);
+
+    // Protocol Administration
+    event TimeoutChanged(uint32 oldTimeout, uint32 newTimeout);
+    event MaxCohortSizeChanged(uint16 oldSize, uint16 newSize);
+    event DispatcherChanged(address oldDispatcher, address newDispatcher);
+
+    struct SigningCohortParticipant {
+        address provider;
+        address operator;
+        bytes signature;
+    }
+
+    struct SigningCohort {
+        address initiator;
+        uint32 initTimestamp;
+        uint32 endTimestamp;
+        // TODO: what's the point of authority if we have a single global cohort?
+        address authority;
+        uint16 totalSignatures;
+        uint16 numSigners;
+        uint16 threshold;
+        SigningCohortParticipant[] signers;
```

Comment:
I remember we had early discussions for the Coordinator about making this an array or a mapping. Each of them have their pros and cons, but I think, in retrospective, I'd have made this a mapping. One of the pros is that this completely removes the need for the whole pre/post sentinel gaps and other concerns that originate from potential changes to the `SigningCohortParticipant` struct.

Reply:
The challenge with using a mapping is that it can't be iterated over. So a function like the following would be difficult:
```solidity
    function getSigners(uint32 cohortId) external view returns (SigningCohortParticipant[] memory) {
         ...
    }
```
Unless, of course, you’ve got some clever solidity workarounds in mind 😉 ?

Reply:
Yep, not a Solidity hack per se, but if we use mappings then there would be no need to implement a `findSigner()` method, at the expense of dynamically constructing a `SigningCohortParticipant[]` array in `getSigners()` by iterating over an array of addresses (which we would have to store separately). So it's really a trade-off that we can discuss, but by no means a required change now.